vm/src/org/mozartoz/truffle/runtime/IdentityPair.java

```diff
@@ -0,0 +1,46 @@
+package org.mozartoz.truffle.runtime;
+
+/**
+ * An identity pair class to be used in general unification and equality
+ * algorithms. Allows recognizing identical pairs of objects within collections.
+ */
+public class IdentityPair implements Comparable<IdentityPair> {
+	private final Object a, b;
+	private final int hashCode;
+
+	public IdentityPair(Object a, Object b) {
+		int ha = System.identityHashCode(a);
+		int hb = System.identityHashCode(b);
+		long hashCode64;
+		if (ha < hb) {
+			this.a = a;
+			this.b = b;
+			hashCode64 = ((long) ha << 32) | hb;
+		} else {
+			this.a = b;
+			this.b = a;
+			hashCode64 = ((long) hb << 32) | ha;
+		}
+		this.hashCode = (int) ((hashCode64 >> 16) ^ hashCode64); // TODO improve distribution?
```

Comment:
Any reason to use this combination?
Maybe use the same mechanism as in Java's `Objects.hash`:
```java
this.hashCode = 31 * ha + hb // if ha < hb
```
or just `ha ^ hb` is maybe good enough.